### PR TITLE
Configuration

### PR DESCRIPTION
**Pull request by constellates**
Update for issue #182 to add configuration tools.

If possible please test this change before merging because there are a lot of changes. Additionally the merge to gh-pages should be merged before this branch to ensure it still functions with these changes.

I've tried to add some validation to the configuration itself but may have missed something so feel free to try bad configurations.

Updates to the README describe the new configuration syntax. Let me know if you have any questions or want any changes to the functionality.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -38,6 +38,81 @@ If you would like to test individual files you can use the file specific checks
...
+
+You can configure the severity of errors by passing a json configuration file with a --c or --config flag to the command line interface or defining a config object on the options object passed during javascript usage.
+
+The basic configuration format is outlined below. All configuration is optional.
+
+    {
+    	"ignore": [],
+    	"warn": [],
+    	"error": [],
+    	ignoredFiles: []
```
Comment (chrisgorgo) [inline README.md]: Missing quotes?

**Comment by constellates** [comment]
I should also mention. I took an initial stab at assigning string keys to each issue so please check if they look reasonable.

**Comment by chrisgorgo** [comment]
Could you add and example "silent" or "info" level issue to this PR. For example lack of _T1w files? This would help me test it in context of BIDS App validator.

**Comment by constellates** [comment]
Alright. I just added a new issue that checks for the presence of T1w scans.

https://github.com/INCF/bids-validator/pull/210/commits/cc20ffe8bc3c29b8f791fd216744fbc5b378b19b#diff-561f5764dd51c29a6809e4f5dd3c98beR255

Its default severity is set to ignore which won't display unless it's configured to, but it will still be silently reported so the web app can store it.

**Comment by chrisgorgo** [comment]
I'm running into problems trying to get this to work. I created a config that should elevate NO_T1W to errors and run it on dataset in which some subjects do not have T1ws:

```
filo@MSI MINGW64 /d/drive/workspace/bids-validator (configuration)
$ bin/bids-validator -c test_config.json /d/example_artifacts_dataset/
        1: You should define 'SliceTiming' for this file. If you don't provide this information slice time correction will not be possible. (code: 13 - SLICE_TIMING_NOT_DEFINED)
                /sub-ds001s07/func/sub-ds001s07_task-balloonanalogrisktask_run-03_bold.nii.gz
                /sub-steph02/func/sub-steph02_task-unknown_bold.nii.gz

        2: Task scans should have a corresponding events.tsv file. If this is a resting state scan you can ignore this warning or rename the task to include the word "rest". (code: 25 - EVENTS_TSV_MISSING)
                /sub-arielle01/func/sub-arielle01_task-unknown_bold.nii.gz
                /sub-ben01/func/sub-ben01_task-unknown_bold.nii.gz
                /sub-ben02/func/sub-ben02_task-unknown_bold.nii.gz
                /sub-ben03/func/sub-ben03_task-unknown_bold.nii.gz
                /sub-ben04/func/sub-ben04_task-unknown_bold.nii.gz
                /sub-ben05/func/sub-ben05_task-unknown_bold.nii.gz
                /sub-ben06/func/sub-ben06_task-unknown_bold.nii.gz
                /sub-ben07/func/sub-ben07_task-unknown_bold.nii.gz
                /sub-ben08/func/sub-ben08_task-unknown_bold.nii.gz
                /sub-ben09/func/sub-ben09_task-unknown_bold.nii.gz
                ... and 2 more files having this issue (Use --verbose to see them all).

        3: Not all subjects contain the same files. Each subject should contain the same number of files with the same naming unless some files are known to be missing. (code: 38 - INCONSISTENT_SUBJECTS)
                /sub-arielle01/anat/sub-arielle01_T1w.nii.gz
                /sub-arielle01/func/sub-arielle01_task-balloonanalogrisktask_run-03_bold.json
                /sub-arielle01/func/sub-arielle01_task-balloonanalogrisktask_run-03_bold.nii.gz
                /sub-ben01/anat/sub-ben01_T1w.nii.gz
                /sub-ben01/func/sub-ben01_task-balloonanalogrisktask_run-03_bold.json
                /sub-ben01/func/sub-ben01_task-balloonanalogrisktask_run-03_bold.nii.gz
                /sub-ben02/anat/sub-ben02_T1w.nii.gz
                /sub-ben02/func/sub-ben02_task-balloonanalogrisktask_run-03_bold.json
                /sub-ben02/func/sub-ben02_task-balloonanalogrisktask_run-03_bold.nii.gz
                /sub-ben03/anat/sub-ben03_T1w.nii.gz
                ... and 62 more files having this issue (Use --verbose to see them all).

        4: Not all subjects/sessions/runs have the same scanning parameters. (code: 39 - INCONSISTENT_PARAMETERS)
                /sub-arielle01/func/sub-arielle01_task-unknown_bold.nii.gz
                /sub-ben01/func/sub-ben01_task-unknown_bold.nii.gz
                /sub-ben09/func/sub-ben09_task-unknown_bold.nii.gz
                /sub-steph01/anat/sub-steph01_T1w.nii.gz
                /sub-steph02/func/sub-steph02_task-unknown_bold.nii.gz

        Summary:                  Available Tasks:                Available Modalities:
        34 Files, 297.59MB        unknown                         bold
        21 - Subjects             Unknown                         T1w
        1 - Session               balloon analog risk task


filo@MSI MINGW64 /d/drive/workspace/bids-validator (configuration)
$ cat test_config.json
{
  "error": ["NO_T1W"]
}

filo@MSI MINGW64 /d/drive/workspace/bids-validator (configuration)
$ ls -al /d/example_artifacts_dataset/
participants.tsv  sub-ben01/        sub-ben03/        sub-ben05/        sub-ben07/        sub-ben09/        sub-kate000/      sub-kate002/      sub-kate006/      sub-kate008/      sub-steph01/
sub-arielle01/    sub-ben02/        sub-ben04/        sub-ben06/        sub-ben08/        sub-ds001s07/     sub-kate001/      sub-kate004/      sub-kate007/      sub-kate013/      sub-steph02/

filo@MSI MINGW64 /d/drive/workspace/bids-validator (configuration)
$ ls -al /d/example_artifacts_dataset/sub-ben01/
total 4
drwxr-xr-x 1 filo 197609 0 Oct  7 16:37 ./
drwxr-xr-x 1 filo 197609 0 Oct 19 13:25 ../
drwxr-xr-x 1 filo 197609 0 Oct  7 16:38 func/
```

As you can see no error was triggered.

**Comment by constellates** [comment]
I think I misunderstood the T1w check. It's currently setup to be triggered if the dataset as a whole does not contain a T1w, but it should be triggered if any one subject doesn't contain a T1w. Is that correct?

**Comment by chrisgorgo** [comment]
Ah yes - it's me who misunderstood the meaning of the error. I can confirm
it works. Triggering it when the there are no T1ws (for any subject) is a
better option in context of validating datasets for apps.

If you could add the error upon lack of T1ws example to the readme that
would be great (in context of app validation). After this small change you
can merge at will.

**Pull request merged**